File: app/src/lib/place-embedded-bid.ts
/**
 * IDL-free encoder + sender for `roundfi-core::place_embedded_bid`
 * (ADR 0012 Fase 2 — lance embutido).
 *
 * The bid moves NO funds: its "price" was already paid through normal
 * `contribute` calls made AHEAD of the pool's clock (ADR 0012 Fase 1
 * prepayment). What this instruction does is swap two entries of the
 * pool's `DrawResult.order` — the bidder's seat takes the CURRENT cycle,
 * and the seat that held it inherits the bidder's original (future)
 * cycle. `order` stays a bijection, so every member is still
 * contemplated exactly once and the payout instructions need no change.
 *
 * Pre-conditions enforced on-chain (mirrored by `lib/lance.ts` so the UI
 * never fires a doomed tx):
 *   - pool.status == Active AND pool.ordering_policy == Sorteio
 *   - !member.defaulted && !member.paid_out
 *   - depth = contributions_paid − current_cycle − 1 >= 1
 *   - depth > pool.current_bid_depth (STRICTLY — ties lose)
 *   - the bidder's own drawn cycle is still in the future
 *
 * Account order MUST match `PlaceEmbeddedBid<'info>` in
 * programs/roundfi-core/src/instructions/place_embedded_bid.rs
 * (5 accounts, no args).
 */

import { Connection, PublicKey, Transaction, TransactionInstruction } from "@solana/web3.js";

import { drawResultPda, memberPda, protocolConfigPda } from "@roundfi/sdk/pda";

import { DEVNET_PROGRAM_IDS } from "./devnet";
import { confirmOrThrow, simulateOrThrow } from "./simulateTx";

// sha256("global:place_embedded_bid")[:8] — precomputed.
//   $ node -e 'console.log(require("crypto").createHash("sha256")
//                .update("global:place_embedded_bid").digest()
//                .subarray(0,8).toString("hex"))'
//   → 74f3af14730d08d6
const PLACE_EMBEDDED_BID_DISCRIMINATOR = Buffer.from([
  0x74, 0xf3, 0xaf, 0x14, 0x73, 0x0d, 0x08, 0xd6,
]);

export interface BuildPlaceEmbeddedBidIxArgs {
  /** Pool PDA — Active sorteio pool (mutable: `current_bid_depth`). */
  pool: PublicKey;
  /** Connected wallet — signs; must own the Member PDA for this pool. */
  memberWallet: PublicKey;
  /** Optional program ID override — for tests against a local program set. */
  programIds?: { core: PublicKey };
}

/** Build the raw `place_embedded_bid()` instruction (no args, 5 accounts). */
export function buildPlaceEmbeddedBidIx(args: BuildPlaceEmbeddedBidIxArgs): TransactionInstruction {
  const core = args.programIds?.core ?? DEVNET_PROGRAM_IDS.core;

  const [config] = protocolConfigPda(core);
  const [member] = memberPda(core, args.pool, args.memberWallet);
  const [draw] = drawResultPda(core, args.pool);

  return new TransactionInstruction({
    programId: core,
    data: PLACE_EMBEDDED_BID_DISCRIMINATOR,
    keys: [
      { pubkey: args.memberWallet, isSigner: true, isWritable: true },
      { pubkey: config, isSigner: false, isWritable: false },
      { pubkey: args.pool, isSigner: false, isWritable: true },
      { pubkey: member, isSigner: false, isWritable: false },
      // The DrawResult is a DECLARED account here (not a remaining one as
      // in claim_payout): the bid's entire effect is the swap it writes.
      { pubkey: draw, isSigner: false, isWritable: true },
    ],
  });
}

export interface SendPlaceEmbeddedBidArgs extends BuildPlaceEmbeddedBidIxArgs {
  connection: Connection;
  /** Wallet adapter callback — usually `wallet.sendTransaction`. */
  sendTransaction: (tx: Transaction, connection: Connection) => Promise<string>;
}

/** Build, simulate, sign-via-wallet and CONFIRM the bid. Returns the signature. */
export async function sendPlaceEmbeddedBid(args: SendPlaceEmbeddedBidArgs): Promise<string> {
  const ix = buildPlaceEmbeddedBidIx(args);
  const tx = new Transaction().add(ix);
  const { blockhash, lastValidBlockHeight } = await args.connection.getLatestBlockhash("confirmed");
  tx.recentBlockhash = blockhash;
  tx.feePayer = args.memberWallet;

  // Dry-run before the wallet signs — never sign a tx that will fail.
  await simulateOrThrow(args.connection, tx);

  const signature = await args.sendTransaction(tx, args.connection);
  // A bid is the one action here that can lose a RACE it passed simulation
  // for (someone deeper landed first, or the contemplated member claimed
  // and the cycle advanced). `confirmTransaction` RESOLVES for a landed-
  // but-reverted tx, so without this the UI would celebrate a bid that
  // never took the slot.
  await confirmOrThrow(args.connection, signature, blockhash, lastValidBlockHeight);
  return signature;
}

File: programs/roundfi-core/src/instructions/place_embedded_bid.rs
//! `place_embedded_bid` — lance embutido (ADR 0012 Phase 2).
//!
//! A sorteio-pool member who has PREPAID installments beyond the pool's
//! current cycle (ADR 0012 Phase 1) may offer that prepayment as a bid to
//! be contemplated NOW: the instruction swaps two entries of the pool's
//! `DrawResult.order` so the bidder's seat takes the CURRENT cycle and the
//! seat that held the current cycle takes the bidder's original (future)
//! drawn cycle.
//!
//! Why a swap is the whole mechanism:
//!   - `order` stays a bijection over `0..members_target`, so EVERY member
//!     is still contemplated exactly once — nobody is skipped, nobody can
//!     be paid twice. The displaced member keeps the same credit, later
//!     (bid beats luck — the consórcio deal).
//!   - `claim_payout` / `crank_payout` / `skip_defaulted_payout` already
//!     read contemplation through `contemplated_cycle_for_seat`, so they
//!     need ZERO changes — the swapped DrawResult IS the new truth, and
//!     the vault waterfall / viability math never notice.
//!   - The bid moves NO funds: the "price" was already paid through the
//!     normal `contribute` split (solidarity/escrow/float preserved).
//!     Pay-after-receiving is untouched — a bid winner contemplated early
//!     still owes every remaining installment, and their prepayment IS
//!     those installments.
//!
//! Within one cycle, bids compete: only a STRICTLY deeper bid (more
//! installments prepaid beyond `current_cycle`) than
//! `pool.current_bid_depth` is accepted; winning re-swaps against the
//! current holder (which may be the previous bidder — the chain of swaps
//! keeps the permutation intact). Cycle advance resets the tracker so the
//! next cycle's competition starts fresh; standing swaps persist in the
//! DrawResult, auditable via the event log.
//!
//! Anti-snipe note (accepted for the devnet canary, flagged in
//! `docs/security/lance-contemplation.md`): a deeper bid can land right
//! before the current holder claims. The claim-vs-bid race is benign —
//! whichever tx lands first wins the slot, and both outcomes are valid
//! permutations. Phase 3 (free bids, external USDC) adds commit-reveal in
//! the #232 mold before it ships.

use anchor_lang::prelude::*;

use crate::constants::{ORDERING_SORTEIO, SEED_CONFIG, SEED_DRAW_RESULT, SEED_MEMBER, SEED_POOL};
use crate::error::RoundfiError;
use crate::state::{DrawResult, Member, Pool, PoolStatus, ProtocolConfig};

#[derive(Accounts)]
pub struct PlaceEmbeddedBid<'info> {
    pub member_wallet: Signer<'info>,

    #[account(
        seeds = [SEED_CONFIG],
        bump = config.bump,
        constraint = !config.paused @ RoundfiError::ProtocolPaused,
    )]
    pub config: Box<Account<'info, ProtocolConfig>>,

    #[account(
        mut,
        seeds = [SEED_POOL, pool.authority.as_ref(), &pool.seed_id.to_le_bytes()],
        bump = pool.bump,
        constraint = pool.status == PoolStatus::Active as u8 @ RoundfiError::PoolNotActive,
        constraint = pool.ordering_policy == ORDERING_SORTEIO @ RoundfiError::EmbeddedBidUnavailable,
    )]
    pub pool: Box<Account<'info, Pool>>,

    #[account(
        seeds = [SEED_MEMBER, pool.key().as_ref(), member_wallet.key().as_ref()],
        bump = member.bump,
        constraint = member.wallet == member_wallet.key() @ RoundfiError::NotAMember,
        constraint = !member.defaulted @ RoundfiError::DefaultedMember,
    )]
    pub member: Box<Account<'info, Member>>,

    #[account(
        mut,
        seeds = [SEED_DRAW_RESULT, pool.key().as_ref()],
        bump = draw.bump,
        constraint = draw.pool == pool.key() @ RoundfiError::InvalidDrawAccount,
    )]
    pub draw: Box<Account<'info, DrawResult>>,
}

pub fn handler(ctx: Context<PlaceEmbeddedBid>) -> Result<()> {
    let pool = &mut ctx.accounts.pool;
    let member = &ctx.accounts.member;
    let draw = &mut ctx.accounts.draw;
    let clock = Clock::get()?;

    // ─── Window: embedded bids close exactly where commits close ────────
    // An embedded bid needs no envelope, so without a window it could be
    // placed *after* watching a sealed bid reveal — and the winning reveal
    // publishes the number to beat (`depth={}` in place_bid_reveal's WON
    // log). Since both paths compete on the same `pool.current_bid_depth`,
    // a member holding prepaid installments could take the cycle with
    // `depth + 1` and no commit, defeating the anti-snipe property the
    // whole sealed-bid phase exists for. `docs/security/lance-contemplation.md`
    // §5.3 claimed that snipe was closed; it was closed only against
    // another SEALED bid (external ADR 0012/0013 audit, M-2).
    //
    // Putting embedded bids on the COMMIT window (`clock < next_cycle_at`)
    // rather than the reveal window is what makes the two paths symmetric:
    // everyone bids blind to the sealed book, then reveals adjudicate
    // against whatever depth is standing. Reusing `BidWindowClosed` keeps
    // the error set unchanged (RoundfiError codes are positional).
    //
    // Embedded-vs-embedded stays openly competitive inside the cycle —
    // that is Phase 2's documented, accepted snipe (§3), now bounded to
    // before any envelope opens instead of running through the auction.
    require!(
        clock.unix_timestamp < pool.next_cycle_at,
        RoundfiError::BidWindowClosed,
    );

    // A member already contemplated has nothing to bid FOR (and letting a
    // paid-out member swap back into the current cycle would be a second
    // payout — the one corruption the bijection alone can't rule out).
    require!(!member.paid_out, RoundfiError::EmbeddedBidUnavailable);

    // The bid metric: installments prepaid BEYOND the one currently due.
    // `contributions_paid == current_cycle + 1` is just being CURRENT (the
    // normal paid-this-cycle state) — bid material starts at the FIRST
    // installment past that (ADR 0012 Phase 1 made it reachable), else any
    // merely-paid-up member could take the cycle with a zero bid.
    let ahead = member.contributions_paid.saturating_sub(pool.current_cycle);
    let depth = ahead.saturating_sub(1);
    require!(depth >= 1, RoundfiError::EmbeddedBidUnavailable);
    require!(depth > pool.current_bid_depth, RoundfiError::EmbeddedBidTooShallow);

    // The bidder's own drawn turn must still be in the FUTURE: equal to the
    // current cycle means they already hold it (no-op), and a past turn
    // implies paid_out (excluded above) — defensive both ways.
    let seat = member.slot_index;
    let my_cycle = draw.cycle_for_seat(seat)?;
    let current = pool.current_cycle;
    require!(my_cycle > current, RoundfiError::EmbeddedBidUnavailable);

    // Find the seat currently holding this cycle. `order` is a bijection
    // over 0..members_target, so exactly one exists; anything else means a
    // corrupted draw account.
    let n = draw.members_target as usize;
    let displaced = draw.order[..n]
        .iter()
        .position(|&c| c == current)
        .ok_or(error!(RoundfiError::InvalidDrawAccount))?;

    // The swap — permutation in, permutation out.
    draw.order[seat as usize] = current;
    draw.order[displaced] = my_cycle;
    pool.current_bid_depth = depth;

    msg!(
        "roundfi-core: embedded bid pool={} cycle={} winner_seat={} depth={} displaced_seat={} displaced_to_cycle={}",
        pool.key(),
        current,
        seat,
        depth,
        displaced,
        my_cycle,
    );
    Ok(())
}
